sql/src/main/java/org/elasticsearch/action/bulk/BulkRetryCoordinator.java

```diff
@@ -21,132 +21,160 @@
-            executor.execute(request, retryBulkActionListener);
-        } else {
-            // new retries will be spawned in new thread because they can block
-            retryExecutorService.schedule(
-                    new Runnable() {
-                        @Override
-                        public void run() {
-                            LOGGER.trace("retry thread [{}] started", Thread.currentThread().getName());
-                            // will block if other retries/writer are active
-                            try {
-                                retryLock.acquireWriteLock();
-                            } catch (InterruptedException e) {
-                                Thread.interrupted();
-                            }
-                            LOGGER.trace("retry thread [{}] executing", Thread.currentThread().getName());
-                            executor.execute(request, retryBulkActionListener);
-                        }
-                    }, currentDelay.getAndAdd(DELAY_INCREMENT), TimeUnit.MILLISECONDS);
+            activeOperations++;
         }
-    }
 
-    private void trace(String message, Object ... args) {
-        if (LOGGER.isTraceEnabled()) {
-            LOGGER.trace("BulkRetryCoordinator: active retries: {} - {}",
-                    retryLock.activeWriters(), String.format(Locale.ENGLISH, message, args));
+        try {
+            retryLock.acquireWriteLock();
+        } catch (InterruptedException e) {
+            operation.responseListener.onFailure(e);
+            return;
         }
+
+        final ActionListener<ShardResponse> triggeringListener = new PendingTriggeringActionListener(operation);
+        operation.executor.execute(operation.request, triggeringListener);
     }
 
-    public void close() {
-        if (!retryExecutorService.isTerminated()) {
-            retryExecutorService.shutdown();
-            try {
-                retryExecutorService.awaitTermination(currentDelay.get() + 100, TimeUnit.MILLISECONDS);
-            } catch (InterruptedException e) {
-                e.printStackTrace();
-            } finally {
-                retryExecutorService.shutdownNow();
+    private void triggerNext() {
+        releaseWriteLock();
+        PendingOperation<ShardRequest, ShardResponse> pendingOperation = null;
+        synchronized (pendingLock) {
+            activeOperations--;
+            if (activeOperations == 0) {
+                pendingOperation = pendingOperations.poll();
             }
         }
-    }
+        trace("triggerNext - pendingOperation=%s pendingOperations=%d", pendingOperation, pendingOperations.size());
 
-    public void shutdown() {
-        retryExecutorService.shutdownNow();
+        if (pendingOperation == null) {
+            return;
+        }
+        retryOperation(pendingOperation);
     }
 
-    private class RetryBulkActionListener implements ActionListener<ShardResponse> {
+    class PendingOperation<Request, Response> {
```

Comment:
it's non-static right now, because it uses the backoff strategy of the BulkRetryCoordinator class
this can be changed though